src/NuGet.Core/NuGet.Protocol.Core.v3/PushCommandResource.cs

```diff
@@ -0,0 +1,232 @@
+        /// <param name="packageSize">The size of package to be pushed.</param>
+        /// <param name="logger">The logger</param>
+        /// <param name="token">The cancellation token</param>
+        public async Task PushPackage(string apiKey,
+            string pathToPackage,
+            long packageSize,
+            ILogger logger,
+            CancellationToken token)
+        {
+            if (_source.IsFile)
+            {
+                PushPackageToFileSystem(pathToPackage);
+            }
+            else
+            {
+                await PushPackageToServer(apiKey, pathToPackage, packageSize, logger, token);
+            }
+        }
+
+        /// <summary>
+        /// Pushes a package to the Http server.
+        /// </summary>
+        /// <param name="apiKey">API key to be used to push the package.</param>
+        /// <param name="pathToPackage">The path of the package to be pushed </param>
+        /// <param name="packageSzie">Size of the package to be pushed.</param>
+        /// <param name="timeout">Time in milliseconds to timeout the server request.</param>
+        /// <param name="logger">The logger</param>
+        /// <param name="token">The cancellation token</param>
+        private async Task PushPackageToServer(
+            string apiKey,
+            string pathToPackage,
+            long packageSzie,
+            ILogger logger,
+            CancellationToken token)
+        {
+            using (var fileStream = new FileStream(pathToPackage, FileMode.Open, FileAccess.Read, FileShare.Read))
+            using (var request = new HttpRequestMessage(HttpMethod.Put, GetServiceEndpointUrl(string.Empty)))
+            {
+                if (!string.IsNullOrEmpty(apiKey))
+                {
+                    request.Headers.Add(ApiKeyHeader, apiKey);
+                }
+
+                using (var content = new MultipartFormDataContent())
+                using (var packageContent = new StreamContent(fileStream))
+                {
+                    packageContent.Headers.ContentType = MediaTypeHeaderValue.Parse("application/octet-stream");
+                    content.Add(packageContent, "package", "package.nupkg");
+                    request.Content = content;
+
+                    var response = await _httpSource.SendAsync(request, logger, token);
+                    response.EnsureSuccessStatusCode();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Pushes a package to a FileSystem.
+        /// </summary>
+        /// <param name="pathToPackage">The path of package to be pushed.</param>
+        private void PushPackageToFileSystem(string pathToPackage)
+        {
+            string root = _source.LocalPath;
+            PackageArchiveReader reader = new PackageArchiveReader(pathToPackage);
+            PackageIdentity packageIdentity = reader.GetIdentity();
+
+            //TODD: support V3 repo style if detect it is
```

Comment:
Likely i will not have time to include V3 support in the next PR, if indeed, i will log a bug